File: internal/indexer/testdata/python/tricky.py
"""Tricky Python fixture for deep attribute analysis."""

import os
import json
from typing import Dict, List, Optional, Any
from collections import defaultdict
from dataclasses import dataclass, field

# --- Shadowing: local shadows global ---

MAX_RETRIES = 5  # module-level constant


def shadow_example():
    MAX_RETRIES = 10  # shadows module-level
    print(MAX_RETRIES)


# --- Multiple inheritance ---


class Serializable:
    def serialize(self) -> str:
        return json.dumps(self.__dict__)


class Printable:
    def display(self) -> None:
        print(str(self))


class Document(Serializable, Printable):
    def __init__(self, title: str, content: str):
        self.title = title
        self.content = content

    def __str__(self) -> str:
        return f"Document({self.title})"


# --- Decorators and properties ---


class Config:
    def __init__(self, data: Dict[str, Any]):
        self._data = data
        self._cache: Dict[str, Any] = {}

    @property
    def host(self) -> str:
        return self._data.get("host", "localhost")

    @host.setter
    def host(self, value: str) -> None:
        self._data["host"] = value

    @staticmethod
    def from_env() -> "Config":
        return Config({"host": os.getenv("HOST", "localhost")})

    @classmethod
    def default(cls) -> "Config":
        return cls({"host": "localhost", "port": 8080})

    # --- Nested function ---
    def process(self, items: List[str]) -> List[str]:
        def transform(item: str) -> str:
            return item.strip().lower()

        return [transform(i) for i in items]


# --- Dataclass ---


@dataclass
class Point:
    x: float
    y: float
    label: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)

    def distance_to(self, other: "Point") -> float:
        return ((self.x - other.x) ** 2 + (self.y - other.y) ** 2) ** 0.5


# --- Builtins usage (should NOT be unresolved external) ---


def use_builtins():
    # builtin functions
    x = len([1, 2, 3])
    y = range(10)
    z = list(y)
    s = str(x)
    i = int("42")
    f = float("3.14")
    b = bool(1)
    d = dict(a=1, b=2)
    t = tuple([1, 2])
    st = set([1, 2, 3])
    _ = isinstance(x, int)
    _ = issubclass(int, object)
    _ = hasattr([], "append")
    _ = getattr([], "append")
    _ = callable(print)
    _ = type(x)
    _ = repr(x)
    _ = hash(42)
    _ = abs(-1)
    _ = min(1, 2)
    _ = max(1, 2)
    _ = sorted([3, 1, 2])
    _ = reversed([1, 2, 3])
    _ = enumerate(z)
    _ = zip([1], [2])
    _ = map(str, [1, 2])
    _ = filter(None, [0, 1])
    _ = any([True, False])
    _ = all([True, True])
    print(s, i, f, b, d, t, st)
    raise ValueError("test")


# --- External stdlib references ---


def use_external_stdlib():
    # os module
    cwd = os.getcwd()
    home = os.environ.get("HOME", "/tmp")

    # json module
    data = json.dumps({"key": "value"})
    parsed = json.loads(data)

    # collections
    counter = defaultdict(int)
    counter["a"] += 1

    return cwd, home, parsed, counter


# --- Name collision: function and class with same prefix ---

Task = None  # module-level variable


class Task:
    """Class with same name as module-level variable."""

    def __init__(self, name: str):
        self.name = name

    def run(self) -> None:
        pass


# --- Dunder methods ---


class Container:
    def __init__(self):
        self._items: List[Any] = []

    def __len__(self) -> int:
        return len(self._items)

    def __getitem__(self, index: int) -> Any:
        return self._items[index]

    def __setitem__(self, index: int, value: Any) -> None:
        self._items[index] = value

    def __contains__(self, item: Any) -> bool:
        return item in self._items

    def __iter__(self):
        return iter(self._items)


# --- Import alias usage ---

dd = defaultdict  # alias for defaultdict

File: internal/indexer/testdata/rust/tricky.rs
//! Tricky Rust fixture for deep attribute analysis.

use std::collections::HashMap;
use std::fmt;
use std::io::{self, Read, Write};

// --- Shadowing: variable shadowing is idiomatic in Rust ---

const MAX_RETRIES: u32 = 5;

fn shadow_example() {
    let max_retries = 10; // shadows constant (different case, but same concept)
    let max_retries = max_retries + 1; // re-shadow within same scope
    println!("{}", max_retries);
}

// --- Trait inheritance ---

trait Serializable {
    fn serialize(&self) -> String;
}

trait Printable: fmt::Display {
    fn display(&self) {
        println!("{}", self);
    }
}

// --- Multiple trait impls ---

struct Document {
    title: String,
    content: String,
}

impl Serializable for Document {
    fn serialize(&self) -> String {
        format!("{{\"title\":\"{}\"}}", self.title)
    }
}

impl fmt::Display for Document {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Document({})", self.title)
    }
}

impl Printable for Document {}

// --- Enum with data ---

enum Shape {
    Circle(f64),
    Rectangle(f64, f64),
    Triangle { base: f64, height: f64 },
}

impl Shape {
    fn area(&self) -> f64 {
        match self {
            Shape::Circle(r) => std::f64::consts::PI * r * r,
            Shape::Rectangle(w, h) => w * h,
            Shape::Triangle { base, height } => 0.5 * base * height,
        }
    }
}

// --- Generic struct + trait bound ---

struct Repository<T: Clone> {
    items: HashMap<u64, T>,
}

impl<T: Clone> Repository<T> {
    fn new() -> Self {
        Repository {
            items: HashMap::new(),
        }
    }

    fn save(&mut self, id: u64, item: T) {
        self.items.insert(id, item);
    }

    fn find(&self, id: u64) -> Option<&T> {
        self.items.get(&id)
    }
}

// --- Builtins / prelude usage (should NOT be unresolved external) ---

fn use_builtins() {
    // Prelude types
    let _opt: Option<i32> = Some(42);
    let _res: Result<i32, String> = Ok(42);
    let _v: Vec<i32> = vec![1, 2, 3];
    let _s: String = String::from("hello");
    let _b: Box<i32> = Box::new(42);

    // Common traits used implicitly
    let _cloned = _v.clone();
    let _debug = format!("{:?}", _v);
    let _display = _s.to_string();

    // Iterator methods
    let _sum: i32 = vec![1, 2, 3].iter().sum();
    let _mapped: Vec<i32> = vec![1, 2, 3].iter().map(|x| x * 2).collect();
    let _filtered: Vec<&i32> = vec![1, 2, 3].iter().filter(|&&x| x > 1).collect();

    // println! is a macro, not a function — but fmt is used
    println!("builtins test: {:?}", _opt);
}

// --- External stdlib references ---

fn use_external_stdlib() -> io::Result<()> {
    // std::collections::HashMap
    let mut map: HashMap<String, i32> = HashMap::new();
    map.insert("key".to_string(), 42);

    // std::io
    let mut buffer = Vec::new();
    let _ = io::stdout().write_all(b"hello");
    let _ = io::stdin().read_to_end(&mut buffer);

    Ok(())
}

// --- Type alias ---

type UserId = u64;
type UserMap = HashMap<UserId, Document>;

// --- Struct with lifetime ---

struct Borrowed<'a> {
    data: &'a str,
}

impl<'a> Borrowed<'a> {
    fn new(data: &'a str) -> Self {
        Borrowed { data }
    }

    fn get(&self) -> &str {
        self.data
    }
}

// --- Closure / function pointer ---

type Transformer = fn(i32) -> i32;

fn apply(f: Transformer, val: i32) -> i32 {
    f(val)
}

fn double(x: i32) -> i32 {
    x * 2
}

// --- Static and const ---

static APP_NAME: &str = "TrickyApp";
const VERSION: u32 = 1;

fn main() {
    shadow_example();

    let doc = Document {
        title: "Hello".to_string(),
        content: "World".to_string(),
    };
    println!("{}", doc.serialize());
    doc.display();

    let circle = Shape::Circle(5.0);
    println!("Area: {}", circle.area());

    let mut repo = Repository::new();
    repo.save(1, doc.title.clone());

    use_builtins();
    let _ = use_external_stdlib();

    let result = apply(double, 21);
    println!("{} v{}: {}", APP_NAME, VERSION, result);
}
